HTML5 beginGradientFill() -> Support for SpreadMethod.REPEAT & SpreadMethod.REFLECT

scanline opened the pull request:
In it's current state, OpenFL ignores the `spreadMethod` parameter passed to the `beginGradientFill()` method when
targeting HTML5. This parameter controls the look of gradients.
Hence the following sample snippet:
```
package;

import openfl.display.Sprite;
import openfl.display.Shape;
import openfl.display.Sprite;
import openfl.display.GradientType;
import openfl.display.SpreadMethod;
import openfl.geom.Matrix;

class Main extends Sprite
{

	public function new()
	{
		super();
		var gradient:Shape;
		var fillType:String = GradientType.LINEAR;
		var colors:Array<UInt> = [0x2f252c, 0xd3ccb2, 0x6e6751];
		var alphas:Array<Float> = [1, 1, 1];
		var ratios:Array<Int> = [80, 150, 255];
		var mat:Matrix = new Matrix();
		var spreadMethod:String = "";
		mat.createGradientBox(150, 100, Math.PI / 4, 0, 0);
		for (a in 0...3)
		{
			switch (a)
			{
				case 0:
					spreadMethod = SpreadMethod.PAD;
				case 1:
					spreadMethod = SpreadMethod.REPEAT;
				case 2:
					spreadMethod = SpreadMethod.REFLECT;
			}
			gradient = new Shape();
			gradient.graphics.beginGradientFill(fillType, colors, alphas, ratios, mat, spreadMethod);
			gradient.graphics.drawRect(0, 0, 250, 250);
			gradient.graphics.endFill();
			gradient.x = a * 260;
			addChild(gradient);
		}
	}

}
```
which the Flashplayer displays like this:
![flashGradient](https://github.com/openfl/openfl/assets/22029272/bba8c196-946f-4bc0-9021-de0fe6ec3a1c)

looks like this when viewed inside the browser:
![openflGradient](https://github.com/openfl/openfl/assets/22029272/a4819b6b-1aca-4d98-befb-b210f74916dd)

With this pull request, things will look way closer to the original:
![fixedGradient](https://github.com/openfl/openfl/assets/22029272/c1c90123-f38e-4cfd-9290-ef042d734185)

Please note though that this just fixes **linear** gradients - I did not put my hands on radial gradients yet.

scanline committed:
Update CanvasGraphics.hx

Support for SpreadMethod.REPEAT & SpreadMethod.REFLECT

joshtynjala commented:
Does the same issue affect native targets, like windows/mac/linux, neko, and hl?

scanline committed:
Update CanvasGraphics.hx

scanline commented:
Actually I never tried to target any of the native targets Josh - so I can't say for sure but gradients have been a big problem of OpenFL ever since.

joshtynjala commented:
Okay, thanks. Well, it's probably safe to assume that SpreadMethod is not implemented on native targets either. I don't see why we'd skip one, and not the other, because Canvas and Cairo have such similar APIs that porting is pretty easy.

I'll be happy to merge this, but it probably won't happen until I or another contributor have time to copy your changes to CairoGraphics.

joshtynjala committed:
Graphics: fix readGraphicsData() returning incorrect caps/joints in line style (closes #2659)

The null default values from the lineStyle() function signature were propagating and being converted to 0, which changed the caps/joints style when running drawGraphicsData() with the result of readGraphicsData()

joshtynjala committed:
CanvasGraphics: restore optimized SpreadMethod.PAD

No need for a second canvas for this one

Also, ran the formatter

dimensionscape committed:
Incl option to skip text cache

Since the measurement cache can be problematic in some use cases, I decided to include the `skip_measurement_cache` flag in the internals. This define is likely to become obsolete and removed in future versions.

dimensionscape committed:
TextEngine: Remove extra return

Woops

scanline committed:
Merge branch 'openfl:develop' into openfl9

joshtynjala commented:
Closed in 692bfdb2521fada232152652c31d760a5dbc5407 (9.4.0-Dev branch)
